src/modules/types.rs

```diff
@@ -14,14 +14,36 @@ pub enum Type {
     Generic
 }
 
+impl Type {
+    pub fn is_subset_of(&self, other: &Type) -> bool {
+        match (self, other) {
+            (Type::Array(left), Type::Array(right)) => {
+                **left != Type::Generic && **right == Type::Generic
+            }
+            (left, Type::Failable(right)) if !matches!(left, Type::Failable(_)) => {
+                *left == **right || left.is_subset_of(right)
```

Comment:
So it's the same as:
```suggestion
                left.is_allowed_in(*right)
```